Assets/Scripts/PathfindingGrid.cs

```diff
@@ -0,0 +1,156 @@
+using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+
+public class PathfindingGrid
+{
+
+    private float cellSize;
+    private int gridWidth;
+    private int gridHeight;
+    private float originX;
+    private float originY;
+
+    private Dictionary<Vector2,List<Vector2>> pathfindingGraph;
+
+    private Dictionary<Vector2,bool> obstacleDict;
+
+    public PathfindingGrid(int width, int height, float cellSize, float gridOriginX, float gridOriginY){
+        this.cellSize = cellSize;
+        this.gridHeight = height;
+        this.gridWidth = width;
+
+        this.originX = gridOriginX;
+        this.originY = gridOriginY;
+
+        this.pathfindingGraph = new Dictionary<Vector2,List<Vector2>>();
+        this.obstacleDict = new Dictionary<Vector2,bool>();
+        MakeObstacleDict();
+        MakeGraph();
+        DebugObstacleDetection();
+        
+    }
+
+    //-----PUBLIC METHODS-----------------------------------------------
+
+    public Vector2 GetCellCenter(int x, int y){
+        Vector2 cellOrigin = (new Vector2(x,y) * this.cellSize) + new Vector2(this.originX,this.originY);
+        return cellOrigin + new Vector2(this.cellSize/2, this.cellSize/2);
+    }
+
+    public Vector2 GetCellFromWorldPosition(Vector3 worldPosition){
+        int x = Mathf.FloorToInt((worldPosition.x - this.originX)/this.cellSize);
+        int y = Mathf.FloorToInt((worldPosition.y - this.originY)/this.cellSize);
+        return new Vector2(x,y);
+    }
+
+    public Vector2 GetClosestCell(Vector3 worldPosition){
+        Vector2 originCell = GetCellFromWorldPosition(worldPosition);
+        if (this.pathfindingGraph.ContainsKey(originCell) == false){
```

Comment:
Small thing but this  `== false` is unnecessary as the `ContainsKey` method returns a boolean value so you can just use this directly by doing `!this.pathfindingGraph.ContainsKey(originCell)`